File: packages/swift/Sources/EcollectSDK/Modules/PaymentSystemsModule.swift
import Foundation

/// Retrieves the payment methods (medios de pago) enabled for this merchant.
public final class PaymentSystemsModule {
    private let config: EcollectConfig
    private let http: HttpClient
    private let session: SessionModule

    init(config: EcollectConfig, http: HttpClient, session: SessionModule) {
        self.config = config
        self.http = http
        self.session = session
    }

    /// Returns all payment systems enabled for this merchant.
    /// - Returns: Array of `PaymentSystemType` describing available payment methods.
    public func getPaymentSystems() async throws -> [PaymentSystemType] {
        let token = try await session.getSessionToken()
        let body = GetPaymentSystemRequest(
            EntityCode: config.entityCode,
            SessionToken: token
        )
        let response: GetPaymentSystemResponse = try await http.post(
            endpoint: "getPaymentSystem",
            body: body
        )
        switch response.ReturnCode {
        case "SUCCESS":
            return response.PaymentSystemArray ?? []
        case "NO_RECORDS":
            return []
        default:
            throw EcollectError.from(returnCode: response.ReturnCode)
        }
    }
}

File: packages/kotlin/src/main/kotlin/com/ecollect/sdk/modules/PaymentSystemsModule.kt
package com.ecollect.sdk.modules

import com.ecollect.sdk.Config
import com.ecollect.sdk.exceptions.*
import com.ecollect.sdk.types.*
import com.ecollect.sdk.utils.HttpClient

class PaymentSystemsModule(
    private val config: Config,
    private val httpClient: HttpClient,
    private val sessionModule: SessionModule
) {
    /**
     * Get all payment systems (channels) enabled for this merchant.
     * Use FiImagesArray.FindKeys to detect card brand from the first digits.
     */
    suspend fun getPaymentSystems(): GetPaymentSystemResponse {
        val token = sessionModule.getSessionToken()
        val request = GetPaymentSystemRequest(
            entityCode = config.etyCode,
            sessionToken = token
        )
        val url = "${config.baseUrl}getPaymentSystem"
        val response: GetPaymentSystemResponse = httpClient.post(url, request)
        return when (response.returnCode) {
            "SUCCESS", "NO_RECORDS" -> response
            "FAIL_APIEXPIREDSESSION" -> {
                sessionModule.invalidate()
                val newToken = sessionModule.getSessionToken()
                val retryResponse: GetPaymentSystemResponse = httpClient.post(
                    url,
                    request.copy(sessionToken = newToken)
                )
                retryResponse
            }
            "FAIL_INVALIDENTITYCODE" -> throw InvalidConfigException("EntityCode is invalid")
            "FAIL_ACCESSDENIED" -> throw AuthenticationException("Session token authentication failed")
            "FAIL_SYSTEM" -> throw NetworkRetryableException("ecollect system error fetching payment systems")
            else -> response
        }
    }
}
